src/ucp/core/ucp_request.inl

```diff
@@ -592,6 +592,110 @@ static UCS_F_ALWAYS_INLINE void ucp_request_recv_buffer_dereg(ucp_request_t *req
                              &req->recv.state, req);
 }
 
+/* Copy UCT memory handles from memh to state->contig, according to md_map,
+   and up to UCP_MAX_OP_MDS */
+static UCS_F_ALWAYS_INLINE void
+ucp_request_init_dt_reg_from_memh(ucp_request_t *req, ucp_md_map_t md_map,
+                                  ucp_mem_h memh, ucp_dt_reg_t *dt_reg)
+{
+    ucp_md_index_t md_index, src_memh_index, dst_memh_index;
+
+    ucs_assertv(dt_reg->md_map == 0, "md_map=0x%" PRIx64, dt_reg->md_map);
+    ucs_assert((dt_reg == &req->send.state.dt.dt.contig) ||
+               (dt_reg == &req->recv.state.dt.contig));
+
+    req->flags    |= UCP_REQUEST_FLAG_USER_MEMH;
+    src_memh_index = 0;
+    dst_memh_index = 0;
+    ucs_for_each_bit(md_index, memh->md_map) {
+        if (md_map & UCS_BIT(md_index)) {
+            dt_reg->memh[dst_memh_index++] = memh->uct[src_memh_index];
+            dt_reg->md_map                |= UCS_BIT(md_index);
+            if (dst_memh_index >= UCP_MAX_OP_MDS) {
+                break;
+            }
+        }
+        ++src_memh_index;
+    }
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_request_check_user_memh(ucp_request_t *req, ucp_mem_h memh, void *buffer,
+                            size_t length, ucs_memory_type_t mem_type)
+{
+    if (ENABLE_PARAMS_CHECK &&
+        ((memh == NULL) || (buffer < memh->address) ||
+         (UCS_PTR_BYTE_OFFSET(buffer, length) >
+          UCS_PTR_BYTE_OFFSET(memh->address, memh->length)) ||
+         (memh->mem_type != mem_type))) {
+        ucs_error("req %p: mismatched memory handle [buffer %p length %zu %s]"
+                  " memh %p [address %p length %zu %s]",
+                  req, buffer, length, ucs_memory_type_names[mem_type], memh,
+                  memh->address, memh->length,
+                  ucs_memory_type_names[memh->mem_type]);
+        return UCS_ERR_INVALID_PARAM;
```

Comment:
maybe just drop user memh as well as when not all mds are covered by it?

Reply:
then will need to have this check in release mode as well